// @cyclonedx/cyclonedx-library/dist.node/spec/consts.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpecVersionDict = exports.Spec1dot6 = exports.Spec1dot5 = exports.Spec1dot4 = exports.Spec1dot3 = exports.Spec1dot2 = void 0;
const componentType_1 = require("../enums/componentType");
const externalReferenceType_1 = require("../enums/externalReferenceType");
const hashAlogorithm_1 = require("../enums/hashAlogorithm");
const ratingMethod_1 = require("../enums/vulnerability/ratingMethod");
const _protocol_1 = require("./_protocol");
const enums_1 = require("./enums");
exports.Spec1dot2 = Object.freeze(new _protocol_1._Spec(enums_1.Version.v1dot2, [
    enums_1.Format.XML,
    enums_1.Format.JSON
], [
    componentType_1.ComponentType.Application,
    componentType_1.ComponentType.Framework,
    componentType_1.ComponentType.Library,
    componentType_1.ComponentType.Container,
    componentType_1.ComponentType.OperatingSystem,
    componentType_1.ComponentType.Device,
    componentType_1.ComponentType.Firmware,
    componentType_1.ComponentType.File
], [
    hashAlogorithm_1.HashAlgorithm.MD5,
    hashAlogorithm_1.HashAlgorithm['SHA-1'],
    hashAlogorithm_1.HashAlgorithm['SHA-256'],
    hashAlogorithm_1.HashAlgorithm['SHA-384'],
    hashAlogorithm_1.HashAlgorithm['SHA-512'],
    hashAlogorithm_1.HashAlgorithm['SHA3-256'],
    hashAlogorithm_1.HashAlgorithm['SHA3-384'],
    hashAlogorithm_1.HashAlgorithm['SHA3-512'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-256'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-384'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-512'],
    hashAlogorithm_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    externalReferenceType_1.ExternalReferenceType.VCS,
    externalReferenceType_1.ExternalReferenceType.IssueTracker,
    externalReferenceType_1.ExternalReferenceType.Website,
    externalReferenceType_1.ExternalReferenceType.Advisories,
    externalReferenceType_1.ExternalReferenceType.BOM,
    externalReferenceType_1.ExternalReferenceType.MailingList,
    externalReferenceType_1.ExternalReferenceType.Social,
    externalReferenceType_1.ExternalReferenceType.Chat,
    externalReferenceType_1.ExternalReferenceType.Documentation,
    externalReferenceType_1.ExternalReferenceType.Support,
    externalReferenceType_1.ExternalReferenceType.Distribution,
    externalReferenceType_1.ExternalReferenceType.License,
    externalReferenceType_1.ExternalReferenceType.BuildMeta,
    externalReferenceType_1.ExternalReferenceType.BuildSystem,
    externalReferenceType_1.ExternalReferenceType.Other
], true, false, true, false, false, [], false, false, false, false, false, false, true, false));
exports.Spec1dot3 = Object.freeze(new _protocol_1._Spec(enums_1.Version.v1dot3, [
    enums_1.Format.XML,
    enums_1.Format.JSON
], [
    componentType_1.ComponentType.Application,
    componentType_1.ComponentType.Framework,
    componentType_1.ComponentType.Library,
    componentType_1.ComponentType.Container,
    componentType_1.ComponentType.OperatingSystem,
    componentType_1.ComponentType.Device,
    componentType_1.ComponentType.Firmware,
    componentType_1.ComponentType.File
], [
    hashAlogorithm_1.HashAlgorithm.MD5,
    hashAlogorithm_1.HashAlgorithm['SHA-1'],
    hashAlogorithm_1.HashAlgorithm['SHA-256'],
    hashAlogorithm_1.HashAlgorithm['SHA-384'],
    hashAlogorithm_1.HashAlgorithm['SHA-512'],
    hashAlogorithm_1.HashAlgorithm['SHA3-256'],
    hashAlogorithm_1.HashAlgorithm['SHA3-384'],
    hashAlogorithm_1.HashAlgorithm['SHA3-512'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-256'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-384'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-512'],
    hashAlogorithm_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    externalReferenceType_1.ExternalReferenceType.VCS,
    externalReferenceType_1.ExternalReferenceType.IssueTracker,
    externalReferenceType_1.ExternalReferenceType.Website,
    externalReferenceType_1.ExternalReferenceType.Advisories,
    externalReferenceType_1.ExternalReferenceType.BOM,
    externalReferenceType_1.ExternalReferenceType.MailingList,
    externalReferenceType_1.ExternalReferenceType.Social,
    externalReferenceType_1.ExternalReferenceType.Chat,
    externalReferenceType_1.ExternalReferenceType.Documentation,
    externalReferenceType_1.ExternalReferenceType.Support,
    externalReferenceType_1.ExternalReferenceType.Distribution,
    externalReferenceType_1.ExternalReferenceType.License,
    externalReferenceType_1.ExternalReferenceType.BuildMeta,
    externalReferenceType_1.ExternalReferenceType.BuildSystem,
    externalReferenceType_1.ExternalReferenceType.Other
], true, false, true, true, false, [], true, false, true, true, true, false, true, false));
exports.Spec1dot4 = Object.freeze(new _protocol_1._Spec(enums_1.Version.v1dot4, [
    enums_1.Format.XML,
    enums_1.Format.JSON
], [
    componentType_1.ComponentType.Application,
    componentType_1.ComponentType.Framework,
    componentType_1.ComponentType.Library,
    componentType_1.ComponentType.Container,
    componentType_1.ComponentType.OperatingSystem,
    componentType_1.ComponentType.Device,
    componentType_1.ComponentType.Firmware,
    componentType_1.ComponentType.File
], [
    hashAlogorithm_1.HashAlgorithm.MD5,
    hashAlogorithm_1.HashAlgorithm['SHA-1'],
    hashAlogorithm_1.HashAlgorithm['SHA-256'],
    hashAlogorithm_1.HashAlgorithm['SHA-384'],
    hashAlogorithm_1.HashAlgorithm['SHA-512'],
    hashAlogorithm_1.HashAlgorithm['SHA3-256'],
    hashAlogorithm_1.HashAlgorithm['SHA3-384'],
    hashAlogorithm_1.HashAlgorithm['SHA3-512'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-256'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-384'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-512'],
    hashAlogorithm_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    externalReferenceType_1.ExternalReferenceType.VCS,
    externalReferenceType_1.ExternalReferenceType.IssueTracker,
    externalReferenceType_1.ExternalReferenceType.Website,
    externalReferenceType_1.ExternalReferenceType.Advisories,
    externalReferenceType_1.ExternalReferenceType.BOM,
    externalReferenceType_1.ExternalReferenceType.MailingList,
    externalReferenceType_1.ExternalReferenceType.Social,
    externalReferenceType_1.ExternalReferenceType.Chat,
    externalReferenceType_1.ExternalReferenceType.Documentation,
    externalReferenceType_1.ExternalReferenceType.Support,
    externalReferenceType_1.ExternalReferenceType.Distribution,
    externalReferenceType_1.ExternalReferenceType.License,
    externalReferenceType_1.ExternalReferenceType.BuildMeta,
    externalReferenceType_1.ExternalReferenceType.BuildSystem,
    externalReferenceType_1.ExternalReferenceType.ReleaseNotes,
    externalReferenceType_1.ExternalReferenceType.Other
], true, true, false, true, true, [
    ratingMethod_1.RatingMethod.CVSSv2,
    ratingMethod_1.RatingMethod.CVSSv3,
    ratingMethod_1.RatingMethod.CVSSv31,
    ratingMethod_1.RatingMethod.OWASP,
    ratingMethod_1.RatingMethod.Other
], true, false, true, true, true, false, true, false));
exports.Spec1dot5 = Object.freeze(new _protocol_1._Spec(enums_1.Version.v1dot5, [
    enums_1.Format.XML,
    enums_1.Format.JSON
], [
    componentType_1.ComponentType.Application,
    componentType_1.ComponentType.Framework,
    componentType_1.ComponentType.Library,
    componentType_1.ComponentType.Container,
    componentType_1.ComponentType.Platform,
    componentType_1.ComponentType.OperatingSystem,
    componentType_1.ComponentType.Device,
    componentType_1.ComponentType.DeviceDriver,
    componentType_1.ComponentType.Firmware,
    componentType_1.ComponentType.File,
    componentType_1.ComponentType.MachineLearningModel,
    componentType_1.ComponentType.Data
], [
    hashAlogorithm_1.HashAlgorithm.MD5,
    hashAlogorithm_1.HashAlgorithm['SHA-1'],
    hashAlogorithm_1.HashAlgorithm['SHA-256'],
    hashAlogorithm_1.HashAlgorithm['SHA-384'],
    hashAlogorithm_1.HashAlgorithm['SHA-512'],
    hashAlogorithm_1.HashAlgorithm['SHA3-256'],
    hashAlogorithm_1.HashAlgorithm['SHA3-384'],
    hashAlogorithm_1.HashAlgorithm['SHA3-512'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-256'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-384'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-512'],
    hashAlogorithm_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    externalReferenceType_1.ExternalReferenceType.VCS,
    externalReferenceType_1.ExternalReferenceType.IssueTracker,
    externalReferenceType_1.ExternalReferenceType.Website,
    externalReferenceType_1.ExternalReferenceType.Advisories,
    externalReferenceType_1.ExternalReferenceType.BOM,
    externalReferenceType_1.ExternalReferenceType.MailingList,
    externalReferenceType_1.ExternalReferenceType.Social,
    externalReferenceType_1.ExternalReferenceType.Chat,
    externalReferenceType_1.ExternalReferenceType.Documentation,
    externalReferenceType_1.ExternalReferenceType.Support,
    externalReferenceType_1.ExternalReferenceType.Distribution,
    externalReferenceType_1.ExternalReferenceType.DistributionIntake,
    externalReferenceType_1.ExternalReferenceType.License,
    externalReferenceType_1.ExternalReferenceType.BuildMeta,
    externalReferenceType_1.ExternalReferenceType.BuildSystem,
    externalReferenceType_1.ExternalReferenceType.ReleaseNotes,
    externalReferenceType_1.ExternalReferenceType.SecurityContact,
    externalReferenceType_1.ExternalReferenceType.ModelCard,
    externalReferenceType_1.ExternalReferenceType.Log,
    externalReferenceType_1.ExternalReferenceType.Configuration,
    externalReferenceType_1.ExternalReferenceType.Evidence,
    externalReferenceType_1.ExternalReferenceType.Formulation,
    externalReferenceType_1.ExternalReferenceType.Attestation,
    externalReferenceType_1.ExternalReferenceType.ThreatModel,
    externalReferenceType_1.ExternalReferenceType.AdversaryModel,
    externalReferenceType_1.ExternalReferenceType.RiskAssessment,
    externalReferenceType_1.ExternalReferenceType.VulnerabilityAssertion,
    externalReferenceType_1.ExternalReferenceType.ExploitabilityStatement,
    externalReferenceType_1.ExternalReferenceType.PentestReport,
    externalReferenceType_1.ExternalReferenceType.StaticAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.DynamicAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.RuntimeAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.ComponentAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.MaturityReport,
    externalReferenceType_1.ExternalReferenceType.CertificationReport,
    externalReferenceType_1.ExternalReferenceType.CodifiedInfrastructure,
    externalReferenceType_1.ExternalReferenceType.QualityMetrics,
    externalReferenceType_1.ExternalReferenceType.POAM,
    externalReferenceType_1.ExternalReferenceType.Other
], true, true, false, true, true, [
    ratingMethod_1.RatingMethod.CVSSv2,
    ratingMethod_1.RatingMethod.CVSSv3,
    ratingMethod_1.RatingMethod.CVSSv31,
    ratingMethod_1.RatingMethod.CVSSv4,
    ratingMethod_1.RatingMethod.OWASP,
    ratingMethod_1.RatingMethod.SSVC,
    ratingMethod_1.RatingMethod.Other
], true, true, true, true, true, false, true, true));
exports.Spec1dot6 = Object.freeze(new _protocol_1._Spec(enums_1.Version.v1dot6, [
    enums_1.Format.XML,
    enums_1.Format.JSON
], [
    componentType_1.ComponentType.Application,
    componentType_1.ComponentType.Framework,
    componentType_1.ComponentType.Library,
    componentType_1.ComponentType.Container,
    componentType_1.ComponentType.Platform,
    componentType_1.ComponentType.OperatingSystem,
    componentType_1.ComponentType.Device,
    componentType_1.ComponentType.DeviceDriver,
    componentType_1.ComponentType.Firmware,
    componentType_1.ComponentType.File,
    componentType_1.ComponentType.MachineLearningModel,
    componentType_1.ComponentType.Data,
    componentType_1.ComponentType.CryptographicAsset
], [
    hashAlogorithm_1.HashAlgorithm.MD5,
    hashAlogorithm_1.HashAlgorithm['SHA-1'],
    hashAlogorithm_1.HashAlgorithm['SHA-256'],
    hashAlogorithm_1.HashAlgorithm['SHA-384'],
    hashAlogorithm_1.HashAlgorithm['SHA-512'],
    hashAlogorithm_1.HashAlgorithm['SHA3-256'],
    hashAlogorithm_1.HashAlgorithm['SHA3-384'],
    hashAlogorithm_1.HashAlgorithm['SHA3-512'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-256'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-384'],
    hashAlogorithm_1.HashAlgorithm['BLAKE2b-512'],
    hashAlogorithm_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    externalReferenceType_1.ExternalReferenceType.VCS,
    externalReferenceType_1.ExternalReferenceType.IssueTracker,
    externalReferenceType_1.ExternalReferenceType.Website,
    externalReferenceType_1.ExternalReferenceType.Advisories,
    externalReferenceType_1.ExternalReferenceType.BOM,
    externalReferenceType_1.ExternalReferenceType.MailingList,
    externalReferenceType_1.ExternalReferenceType.Social,
    externalReferenceType_1.ExternalReferenceType.Chat,
    externalReferenceType_1.ExternalReferenceType.Documentation,
    externalReferenceType_1.ExternalReferenceType.Support,
    externalReferenceType_1.ExternalReferenceType.SourceDistribution,
    externalReferenceType_1.ExternalReferenceType.Distribution,
    externalReferenceType_1.ExternalReferenceType.DistributionIntake,
    externalReferenceType_1.ExternalReferenceType.License,
    externalReferenceType_1.ExternalReferenceType.BuildMeta,
    externalReferenceType_1.ExternalReferenceType.BuildSystem,
    externalReferenceType_1.ExternalReferenceType.ReleaseNotes,
    externalReferenceType_1.ExternalReferenceType.SecurityContact,
    externalReferenceType_1.ExternalReferenceType.ModelCard,
    externalReferenceType_1.ExternalReferenceType.Log,
    externalReferenceType_1.ExternalReferenceType.Configuration,
    externalReferenceType_1.ExternalReferenceType.Evidence,
    externalReferenceType_1.ExternalReferenceType.Formulation,
    externalReferenceType_1.ExternalReferenceType.Attestation,
    externalReferenceType_1.ExternalReferenceType.ThreatModel,
    externalReferenceType_1.ExternalReferenceType.AdversaryModel,
    externalReferenceType_1.ExternalReferenceType.RiskAssessment,
    externalReferenceType_1.ExternalReferenceType.VulnerabilityAssertion,
    externalReferenceType_1.ExternalReferenceType.ExploitabilityStatement,
    externalReferenceType_1.ExternalReferenceType.PentestReport,
    externalReferenceType_1.ExternalReferenceType.StaticAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.DynamicAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.RuntimeAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.ComponentAnalysisReport,
    externalReferenceType_1.ExternalReferenceType.MaturityReport,
    externalReferenceType_1.ExternalReferenceType.CertificationReport,
    externalReferenceType_1.ExternalReferenceType.CodifiedInfrastructure,
    externalReferenceType_1.ExternalReferenceType.QualityMetrics,
    externalReferenceType_1.ExternalReferenceType.POAM,
    externalReferenceType_1.ExternalReferenceType.ElectronicSignature,
    externalReferenceType_1.ExternalReferenceType.DigitalSignature,
    externalReferenceType_1.ExternalReferenceType.RFC9116,
    externalReferenceType_1.ExternalReferenceType.Other
], true, true, false, true, true, [
    ratingMethod_1.RatingMethod.CVSSv2,
    ratingMethod_1.RatingMethod.CVSSv3,
    ratingMethod_1.RatingMethod.CVSSv31,
    ratingMethod_1.RatingMethod.CVSSv4,
    ratingMethod_1.RatingMethod.OWASP,
    ratingMethod_1.RatingMethod.SSVC,
    ratingMethod_1.RatingMethod.Other
], true, true, true, true, true, true, true, true));
exports.SpecVersionDict = Object.freeze({
    [enums_1.Version.v1dot6]: exports.Spec1dot6,
    [enums_1.Version.v1dot5]: exports.Spec1dot5,
    [enums_1.Version.v1dot4]: exports.Spec1dot4,
    [enums_1.Version.v1dot3]: exports.Spec1dot3,
    [enums_1.Version.v1dot2]: exports.Spec1dot2
});
