// @cyclonedx/cyclonedx-library/dist.node/factories/license.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicenseFactory = void 0;
const license_1 = require("../models/license");
const spdx_1 = require("../spdx");
class LicenseFactory {
    makeFromString(value) {
        try {
            return this.makeSpdxLicense(value);
        }
        catch {
        }
        try {
            return this.makeExpression(value);
        }
        catch {
        }
        return this.makeNamedLicense(value);
    }
    makeExpression(value) {
        const expression = String(value);
        if ((0, spdx_1.isValidSpdxLicenseExpression)(expression)) {
            return new license_1.LicenseExpression(expression);
        }
        throw new RangeError('Invalid SPDX license expression');
    }
    makeDisjunctive(value) {
        try {
            return this.makeSpdxLicense(value);
        }
        catch {
            return this.makeNamedLicense(value);
        }
    }
    makeSpdxLicense(value) {
        const fixed = (0, spdx_1.fixupSpdxId)(String(value));
        if (undefined === fixed) {
            throw new RangeError('Unsupported SPDX license ID');
        }
        return new license_1.SpdxLicense(fixed);
    }
    makeNamedLicense(value) {
        return new license_1.NamedLicense(String(value));
    }
}
exports.LicenseFactory = LicenseFactory;
